sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala

```diff
@@ -381,6 +382,64 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
     }
   }
 
+  private def prunePartitionSchema(partSchema: StructType, projectList: Seq[Expression],
+      filterCondition: Option[Expression] = None): StructType = {
+    // Luckily partition columns do not support nested types. So only need to prune the
+    // top level columns.
+    StructType(partSchema.filter(f =>
+      // The partition column used either in the projectList or the filterCondition will
+      // be kept.
+      (projectList ++ filterCondition).exists { expr =>
+        expr.find {
+          // It is safe to use the column name for comparison for the two cases listed in
+          // `prunePartitionForFileSourceScan`
+          case attr: AttributeReference => attr.name == f.name
+          case _ => false
+        }.isDefined
+      }))
+  }
+
+  // This tries to prune the partition schema for GpuFileSourceScanExec by leveraging
+  // the project list of the first GpuProjectExec after a GpuFileSourceScanExec.
+  private def prunePartitionForFileSourceScan(plan: SparkPlan): SparkPlan = plan match {
+    case p @ GpuProjectExec(projectList, fss: GpuFileSourceScanExec, _) =>
+      // A ProjectExec next to FileSourceScanExec, for cases like
+      //   df.groupBy("b").agg(max($"a")), or
+      //   df.select("b")
+      val prunedPartSchema = prunePartitionSchema(fss.relation.partitionSchema, projectList)
+      p.withNewChildren(Seq(
+        fss.copy(requiredPartitionSchema = Some(prunedPartSchema))(fss.rapidsConf)))
```

Comment:
To reduce the amount of copied code, could leverage a utility method within this method, e.g.:
```scala
    def withPrunedPartSchema(
        fss: GpuFileSourceScanExec,
        projectList: Seq[Expression]): GpuFileSourceScanExec = {
      val prunedPartSchema = prunePartitionSchema(fss.relation.partitionSchema, projectList)
      fss.copy(requiredPartitionSchema = Some(prunedPartSchema))(fss.rapidsConf)
    }

[...]
    p.withNewChildren(Seq(withPrunedPartSchema(fss, projectList)))
```

Reply:
Good suggestion, updated